src/main/scala/diplomacy/Nodes.scala

```diff
@@ -253,6 +265,27 @@ sealed abstract class MixedNode[DI, UI, EI, BI <: Data, DO, UO, EO, BO <: Data](
   val inward = this
   val outward = this
 
+  /** debug info of nodes binding. */
+  def bindingInfo: String =
+    s"""${oBindings.size} outward nodes binded: [${oBindings.map(n => s"${n._3.serialize}-${n._2.name}").mkString(",")}]
+       |${iBindings.size} inward nodes binded: [${iBindings.map(n => s"${n._3.serialize}-${n._2.name}").mkString(",")}]
+       |""".stripMargin
+
+  /** debug info of ports connecting. */
+  def connectedPortsInfo: String =
+    s"""${oPorts.size} outward ports connected: [${oPorts.map(_._2.name).mkString(",")}]
+       |${iPorts.size} inward ports connected: [${iPorts.map(_._2.name).mkString(",")}]
+       |""".stripMargin
+
+  /** debug info of parameters propagations. */
+  def parametersInfo: String =
+    s"""${doParams.size} downstream outward parameters: [${doParams.mkString(",")}]
+       |${uoParams.size} upstream outward parameters: [${uoParams.mkString(",")}]
+       |${diParams.size} downstream inward parameters: [${diParams.mkString(",")}]
+       |${uiParams.size} upstream inward parameters: [${uiParams.mkString(",")}]
+       |""".stripMargin
```

Comment:
I am a little worried that for some protocols the amount of parameters printed out here will be very large. For some tilelink nodes you will get a list of every master and slave in the graph, for example! But maybe in the failure cases you really do want as much info as possible.

Reply:
Yes I do think so, trying to print as much as possible information for debugging use.